lib/Sema/CSDiag.cpp

```diff
@@ -5421,8 +5431,92 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
+        auto lookup = CS.lookupMember(baseType->getRValueType(),
+                                      DeclName(DeclBaseName::createSubscript()));
+        if (lookup.empty())
+          return true;
+
+        auto argExpr = typeCheckChildIndependently(callExpr->getArg(), Type(),
+                                                   CTP_CallArgument);
+        if (!argExpr)
+          return true;
+
+        SmallVector<Identifier, 2> scratch;
+        ArrayRef<Identifier> argLabels = callExpr->getArgumentLabels(scratch);
+        SmallVector<OverloadChoice, 2> choices;
+        auto hasTrailClosure = callArgHasTrailingClosure(argExpr);
+
+        for (auto candidate : lookup)
+          choices.push_back(OverloadChoice(baseType, candidate.getValueDecl(),
+                                           UDE->getFunctionRefKind()));
+
+        CalleeCandidateInfo candidates(baseType, choices,
+                                       hasTrailClosure, CS, true);
+
+        auto argType = CS.getType(argExpr);
+        auto params = swift::decomposeArgType(argType, argLabels);
+
+        using Closeness = CalleeCandidateInfo::ClosenessResultTy;
+
+        candidates.filterList([&](UncurriedCandidate candidate) -> Closeness {
+          auto candFuncType = candidate.getUncurriedFunctionType();
+          if (!candFuncType)
+            return {CC_GeneralMismatch, {}};
+
+          auto candParams = candFuncType->getParams();
+
+          if (params.size() != candParams.size())
+            return {CC_GeneralMismatch, {}};
+
+          for (unsigned i = 0, e = params.size(); i < e; i ++) {
+            auto type1 = params[i].getType()->
+              getUnlabeledType(CS.getASTContext());
+            auto type2 = candParams[i].getType()->
+              getUnlabeledType(CS.getASTContext());
+
+            if (type1->isEqual(type2))
+              continue;
+            if (CS.TC.isSubtypeOf(type2, type1, CS.DC))
+              return {CC_GeneralMismatch, {}};
+          }
+          return {CC_ExactMatch, {}};
+        });
+
+        auto *locator = CS.getConstraintLocator(UDE, ConstraintLocator::Member);
+        auto memberRange = baseExpr->getSourceRange();
+        if (locator)
+          locator = simplifyLocator(CS, locator, memberRange);
+
+        auto baseLoc = baseExpr->getLoc();
+        auto nameLoc = DeclNameLoc(memberRange.Start);
+
+        auto diag = diagnose(baseLoc, diag::did_you_mean_subscript_operator);
+
+        if (candidates.closeness != CC_ExactMatch)
+          return true;
+
+        diag.fixItReplace(SourceRange(argExpr->getStartLoc()), "[");
+        diag.fixItRemove(nameLoc.getSourceRange());
+        diag.fixItRemove(SourceRange(UDE->getDotLoc()));
+
+        if (CS.TC.Context.SourceMgr.extractText({argExpr->getEndLoc(), 1}) == ")")
+          diag.fixItReplace(SourceRange(argExpr->getEndLoc()), "]");
+        else
+          diag.fixItInsertAfter(argExpr->getEndLoc(), "]");
+      }
       return true;
```

Comment:
It would be great to change this to return `CS.TC.Diags.hadAnyError();` as well... Usages of `true` like that warrant more audit :(

Reply:
Yes, we do not trust `typeCheckChildIndependently` to always produce a diagnostic, I have recently found places when it doesn’t.